### PR TITLE
Add all reduced functional arguments.

### PR DESCRIPTION
**Pull request by Ig-dolci**
# Description

### CODE REVIEW

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 1]
```diff
@@ -59,8 +80,17 @@ class EnsembleReducedFunctional(ReducedFunctional):
...
+                 gather_functional=None, derivative_components=None,
+                 scale=1.0, tape=None, eval_cb_pre=lambda *args: None,
+                 eval_cb_post=lambda *args: None,
+                 derivative_cb_pre=lambda controls: controls,
+                 derivative_cb_post=lambda checkpoint, derivative_components, controls: derivative_components
+                 ):
+        super(EnsembleReducedFunctional, self).__init__(
+            J, control, derivative_components=derivative_components,
+            scale=scale, tape=tape, eval_cb_pre=eval_cb_pre,
+            eval_cb_post=eval_cb_post, derivative_cb_pre=derivative_cb_pre,
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:92]: Add hessian args

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 2]
```diff
@@ -45,6 +47,25 @@ class EnsembleReducedFunctional(ReducedFunctional):
         ``Ensemble.ensemble comm``.
     gather_functional : An instance of the :class:`pyadjoint.ReducedFunctional`.
         that takes in all of the Js.
+    derivative_components : list of int
+        The indices of the controls that the derivative should be computed with respect to.
+        If present, it overwrites ``derivative_cb_pre`` and ``derivative_cb_post``.
+    scale : float
+        A scaling factor applied to the functional and its gradient(with respect to the control).
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:54]:
```suggestion
        A scaling factor applied to the functional and its gradient (with respect to the control).
```

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 3]
```diff
@@ -45,6 +47,25 @@ class EnsembleReducedFunctional(ReducedFunctional):
...
+    tape : pyadjoint.Tape
+        A tape object that the reduced functional will use to evaluate the functional and
+        its gradients (or derivatives).
+    eval_cb_pre : :func:
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:58]: Check if `eval_cb_pre` is in the code.

File: firedrake/adjoint/ensemble_reduced_functional.py (hunk 1 above)
Comment (Ig-dolci) [reply firedrake/adjoint/ensemble_reduced_functional.py:99]: done

File: firedrake/adjoint/ensemble_reduced_functional.py (hunk 3 above)
Comment (Ig-dolci) [reply firedrake/adjoint/ensemble_reduced_functional.py:58]: done

**Review by dham — APPROVED** [review]

**Pull request merged**